### DISCUSSION TITLE
be开发环境调试，有些行无法打断点

### DISCUSSION THREAD

**Discussion by stgztsw** [apache/doris#7666] [A - General / Q&A] [answered]
本人是c++小白，想调试一下doris的be，最近抽时间在linux上搞了一个开发环境，版本是0.15branch，一路编译通过，开始调试。但是单步调试的时候有些代码行会被跳过，并且无法在这些行打上断点(断点打在这些行上会跳到其他行)，比如doris_main.cpp的main函数下面两行代码。
 string pid_file = string(getenv("PID_DIR")) + "/be.pid";
 int fd = open(pid_file.c_str(), O_RDWR | O_CREAT, S_IRUSR | S_IWUSR | S_IRGRP | S_IWGRP);
但实际这些代码是被执行了，因为变量监控这边能看到这些变量被赋值了。
一开始以为是调试代码和执行代码不一致，所以在main函数里面加了大量的printf日志输出，发现其实代码是一致的。
请问大家有类似的问题吗？哪位大神知道原因吗?

**Comment by BiteTheDDDDt** [comment, marked as answer]
因为cpp程序经过编译器优化, 实际执行的代码已经不是你看到的那样了. 
可以试着开启debug模式编译, 这会降低编译优化等级, 不过也仍然会有不能打断点的地方.

**Reply by stgztsw** [reply]
哦哦，是这样的啊，谢了。我试一下啊

**Comment by Cherry-Puppy** [comment]
@stgztsw @BiteTheDDDDt 请问be是怎么调试的呢，我这边用 clion attach be的进程 
![image](https://user-images.githubusercontent.com/46736583/172605691-da119274-af3b-4067-9e27-aee590814950.png)
但是打断点没用，
![image](https://user-images.githubusercontent.com/46736583/172605756-335c5d3d-62fe-4bb4-a217-7901b3024f94.png)
build 我是用的 debug 模式编译的，求指教

**Reply by BiteTheDDDDt** [reply]
有的地方可能就是打不了断点，你可以试试打在附近，或者打log

**Reply by Cherry-Puppy** [reply]
![image](https://user-images.githubusercontent.com/46736583/172607197-d6028a54-6819-4120-994c-aad7c8c76ddc.png)
主要是没有任何的变量输出，按道理是有的吧 @BiteTheDDDDt 麻烦指导一下 谢谢

**Reply by BiteTheDDDDt** [reply]
你设置对你的二进制文件目录了吗,`"${workspaceFolder}/output/be/lib/palo_be"`之类的,我不了解clion是怎么设置的,不过应该还是需要这个路径

**Reply by Cherry-Puppy** [reply]
![image](https://user-images.githubusercontent.com/46736583/172611643-98cec062-b784-4103-b38c-faf0a4d5afbe.png)
我试了下 vs code 就一直在加载调试中。

**Reply by BiteTheDDDDt** [reply]
看一下配置文件？

**Reply by Cherry-Puppy** [reply]
{
    "version": "0.2.0",
    "configurations": [
        {
            "name": "(gdb) Attach",
            "type": "cppdbg",
            "request": "attach",
            "program": "/code/incubator-doris/output/be/lib/palo_be",
            "processId": 23252,
            "MIMode": "gdb",
            "setupCommands": [
                {
                    "description": "Enable pretty-printing for gdb",
                    "text": "-enable-pretty-printing",
                    "ignoreFailures": true
                }
            ]
        },
        {
            "name": "(gdb) Launch",
            "type": "cppdbg",
            "request": "launch",
            "program": "/code/incubator-doris/output/be/lib/palo_be",
            "args": [],
            "stopAtEntry": false,
            "cwd": "/code/incubator-doris/output/be",
            "environment": [
                {
                    "name": "DORIS_HOME",
                    "value": "/code/incubator-doris/output/be"
                },
                {
                    "name": "UDF_RUNTIME_DIR",
                    "value": "/code/incubator-doris/output/be/lib/udf-runtime"
                },
                {
                    "name": "LOG_DIR",
                    "value": "/code/incubator-doris/output/be/log"
                },
                {
                    "name": "PID_DIR",
                    "value": "/code/incubator-doris/output/be/bin"
                }
            ],
            "externalConsole": false,
            "MIMode": "gdb",
            "setupCommands": [
                {
                    "description": "Enable pretty-printing for gdb",
                    "text": "-enable-pretty-printing",
                    "ignoreFailures": true
                }
            ]
        }
    ]
}

**Reply by BiteTheDDDDt** [reply]
看了下好像没啥问题,你确定机器上有gdb吗

**Reply by BiteTheDDDDt** [reply]
一般来说，有正确安装gdb，pid和二进制path都写对应该就行了，gdb加载会比较慢。或者也可以试试lldb

**Reply by Cherry-Puppy** [reply]
![image](https://user-images.githubusercontent.com/46736583/172613076-78ee5ba2-081d-40b4-87da-e2713483e137.png)
有的，是还需要 gdbserver 挂一个端口 绑定 be 的 pid 吗？

**Reply by Cherry-Puppy** [reply]
lldb 是怎么配置的呢？

**Reply by BiteTheDDDDt** [reply]
{
    // 使用 IntelliSense 了解相关属性。 
    // 悬停以查看现有属性的描述。
    // 欲了解更多信息，请访问: https://go.microsoft.com/fwlink/?linkid=830387
    "version": "0.2.0",
    "configurations": [
        {
            "type": "lldb",
            "request": "attach",
            "pid": "${command:pickProcess}",
            "name": "Debug",
            "program": "${workspaceFolder}/output/be/lib/palo_be"
        }
    ]
}

装好lldb(ldb-toolchain里也有),vsc上安装`CodeLLDB`插件,然后参考我这个配下config就行

**Reply by Cherry-Puppy** [reply]
![image](https://user-images.githubusercontent.com/46736583/172616300-32d5ab42-62f6-4098-8c6d-96488cf0e902.png)
貌似会出现这样的问题
![image](https://user-images.githubusercontent.com/46736583/172616459-352e3e6e-e0fd-4045-82e6-2a775040429f.png)

**Reply by BiteTheDDDDt** [reply]
> ![image](https://user-images.githubusercontent.com/46736583/172616300-32d5ab42-62f6-4098-8c6d-96488cf0e902.png) 貌似会出现这样的问题 ![image](https://user-images.githubusercontent.com/46736583/172616459-352e3e6e-e0fd-4045-82e6-2a775040429f.png)

glibc版本太低...没啥好办法

**Reply by Cherry-Puppy** [reply]
@BiteTheDDDDt hello 又来打扰了，
![image](https://user-images.githubusercontent.com/46736583/173730005-28d2860e-dc08-4cac-af25-d9728f6588aa.png)
我直接 gdb attach pid 就退出访问非法内存了
![image](https://user-images.githubusercontent.com/46736583/173730105-94f7427e-b151-4194-8dbd-cb0f30354008.png)
请问大家是怎么gdb调试的呢？？？
